File: runCompactness.m
% Copyright (c) 2017, Jose Dolz .All rights reserved.
% 
% Redistribution and use in source and binary forms, with or without modification,
% are permitted provided that the following conditions are met:
% 
%     1. Redistributions of source code must retain the above copyright notice,
%        this list of conditions and the following disclaimer.
%     2. Redistributions in binary form must reproduce the above copyright notice,
%        this list of conditions and the following disclaimer in the documentation
%        and/or other materials provided with the distribution.
% 
%     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
%     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
%     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
%     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
%     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
%     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
%     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
%     OTHER DEALINGS IN THE SOFTWARE.
% 
% Jose Dolz. Dec, 2017.
% email: jose.dolz.upv@gmail.com
% LIVIA Department, ETS, Montreal.

rng(1,'twister');

clc
clear all
close all


%% Load data
global AORTA_MRI; AORTA_MRI = 1;
global ESOPHAGUS_CT; ESOPHAGUS_CT = 2;
global RIGHTVENT_MRI; RIGHTVENT_MRI = 3;
global AORTA_CT; AORTA_CT = 4;

example = RIGHTVENT_MRI;

switch example
    case 1
        disp(' Running Aorta MRI example...');
        % Load data
        load('Data/mriAorta.mat'); % CT image
        load('Data/gtMRIAorta.mat'); % Ground truth
        load('Data/probMapMRIAorta.mat'); % CNN result (prob Map)
        img = mri;

        % General parameters
        ParamsADMM.sigma = 25; 
        % Specific ADMM parameters
        ParamsADMM.lambda = 5000; % Compactness prior 
        ParamsADMM.lambda0 = 0.5; % MRF binary potential 
        ParamsADMM.mu1 = 2000;
    case 2
        disp(' Running Esophagus CT example...');
        % Load data
        load('Data/ctEsophagus.mat'); % CT image
        load('Data/gtEsophagus.mat'); % Ground truth
        load('Data/probMapEsophagus.mat'); % CNN result (prob Map)
        img = ct;

        % General parameters
        ParamsADMM.sigma = 1000; 
        % Specific ADMM parameters
        ParamsADMM.lambda = 1000; % Compactness prior 
        ParamsADMM.lambda0 = 0.05; % MRF binary potential
        ParamsADMM.mu1 = 2000;
    case 3
        disp(' Running Right Ventricle MRI example...');
        % Load data
        load('Data/mriRV.mat'); % CT image
        load('Data/gtRV.mat'); % Ground truth
        load('Data/probMapRV.mat'); % CNN result (prob Map)
        img = mri;

        % General parameters
        ParamsADMM.sigma = 100;  
        % Specific ADMM parameters
        ParamsADMM.lambda = 20000; % Compactness prior 
        ParamsADMM.lambda0 = 0.5; % MRF binary potential 
        ParamsADMM.mu1 = 5000;
    otherwise
        disp(' Running Aorta CT example...');
        % Load data
        load('Data/ctAorta.mat'); % CT image
        load('Data/gtCTAorta.mat'); % Ground truth
        load('Data/probMapCTAorta.mat'); % CNN result (prob Map)
        img = ct;

        % General parameters
        ParamsADMM.sigma = 1000;  
        % Specific ADMM parameters
        ParamsADMM.lambda = 3000; % Compactness prior 
        ParamsADMM.lambda0 = 0.5; % MRF binary potential
        ParamsADMM.mu1 = 2000;
end


%% --------- Common ADMM Compactness PARAMETERS ------------------------
%% Parameters
% General parameters
ParamsADMM.imageScale = 1;
ParamsADMM.noise = 8;

kernelSize = 3;
ParamsADMM.Kernel = ones(kernelSize);
ParamsADMM.Kernel((kernelSize+1)/2,(kernelSize+1)/2) = 0;
ParamsADMM.eps = 1e-10;

% Method parameters (Common to all four applications)
ParamsADMM.mu2 = 50; 
ParamsADMM.mu1Fact = 1.01; % Set between 1 and 1.01 
ParamsADMM.mu2Fact = 1.01; % Set between 1 and 1.01 

ParamsADMM.solvePCG = true; % Use pre-conditioned CG algorithm
ParamsADMM.maxLoops = 1000; % Number of iterations

% Display options
ParamsADMM.dispSeg = false;
ParamsADMM.dispCost = false;

options.display = true;


%% Run compactness
% Load BK library
BK_LoadLib;

% Threshold the probability map from the CNN
% NOTE: This probability map can come from any other model (i.e. loglikelihood, chan-vese,etc...)
CNNSeg = zeros(size(probMap));        
CNNSeg(find(probMap>=0.5)) = 1;

ParamsADMM.GroundTruth = gt;

% Segment image
% NOTE: This function will compute segmentation by employing our
% compactness term (SegADMM) and serial BK algorithm (SegGCs)
disp('Starting compactness segmentation...');
[SegADMM,SegGCs,res] = compactnessSegProbMap(img, probMap, ParamsADMM);

% Evaluate results
[diceADMM,precisionADMM,recallADMM] = evalResults(logical(SegADMM), ParamsADMM.GroundTruth);
[diceGCs,precisionGCs,recallGCs]    = evalResults(logical(SegGCs),  ParamsADMM.GroundTruth);
[diceCNN,precisionCNN,recallCNN]    = evalResults(logical(CNNSeg),  ParamsADMM.GroundTruth);

disp(['CNN: ', num2str(diceCNN), ' vs. (CNN + GCs): ', num2str(diceGCs), ' vs. (CNN+GCs+Comp): ',num2str(diceADMM)]);

if options.display
    drawResults(img, gt, CNNSeg, SegGCs, SegADMM); 
end

File: python/runCompactness.py
#!/usr/bin/env python3

import numpy as np
import scipy as sp
import scipy.io
import matplotlib.pyplot as plt
from sys import argv

from ADMM import compactness_seg_prob_map, Params


def wrap_load(name, path):
    return sp.io.loadmat(path)[name]


def load_and_config(choice):
    params = Params()
    params._v = True

    print(choice)
    if choice == "RIGHTVENT_MRI":
        img = wrap_load('mri', '../Data/mriRV.mat')
        grd_truth = wrap_load('gt', '../Data/gtRV.mat')
        probMap = wrap_load('probMap', '../Data/probMapRV.mat')

        params._sigma = 100
        params._lambda = 20000
        params._lambda0 = .5
        params._mu1 = 5000
    elif choice == "AORTA_MRI":
        img = wrap_load('mri', '../Data/mriAorta.mat')
        grd_truth = wrap_load('gt', '../Data/gtMRIAorta.mat')
        probMap = wrap_load('probMap', '../Data/probMapMRIAorta.mat')

        params._sigma = 25
        params._lambda = 5000
        params._lambda0 = .5
        params._mu1 = 2000
    elif choice == "ESOPHAGUS_CT":
        img = wrap_load('ct', '../Data/ctEsophagus.mat')
        grd_truth = wrap_load('gt', '../Data/gtEsophagus.mat')
        probMap = wrap_load('probMap', '../Data/probMapEsophagus.mat')

        params._sigma = 1000
        params._lambda = 1000
        params._lambda0 = .5
        params._mu1 = 2000
    elif choice == "AORTA_CT":
        img = wrap_load('ct', '../Data/ctAorta.mat')
        grd_truth = wrap_load('gt', '../Data/gtCTAorta.mat')
        probMap = wrap_load('probMap', '../Data/probMapCTAorta.mat')

        params._sigma = 1000
        params._lambda = 3000
        params._lambda0 = .5
        params._mu1 = 2000
    else:
        raise NameError("{} is not a valid choice".format(choice))

    return img, grd_truth, probMap, params


def eval_results(seg, ground):
    t_p = np.sum(seg & ground)  # Sum works because those are booleans
    p_s = np.sum(seg)
    p_g = np.sum(ground)

    dice_index = (2 * t_p) / (p_s + p_g)
    precision = t_p / p_s
    recall = t_p / p_g

    return dice_index, precision, recall


def draw_results(img, grd_truth, segCNN, segGCs, segADMM):
    fig, axes = plt.subplots(nrows=1, ncols=4)

    figs = [(grd_truth, "Ground Truth"),
            (segCNN, "seg (CNN)"),
            (segGCs, "Seg (Gcs)"),
            (segADMM, "Seg (ADMM)")]

    for axe, fig in zip(axes.flat, figs):
        axe.imshow(img, cmap="Greys")
        axe.set_title(fig[1])
        axe.contour(fig[0])

    plt.show()


if __name__ == "__main__":
    choice = "RIGHTVENT_MRI"
    if len(argv) > 1:
        choice = argv[1]

    img, grd_truth, probMap, params = load_and_config(choice)

    params._GC = False
    # if not params._GC:
    params._lambda /= 50
    params._mu1 /= 200
    params._mu2 /= 1000
    # params._lambda = 200
    # params._mu1 = 50

    segCNN = probMap >= 0.5

    print("Starting compactness segmentation...")
    segADMM, segGCs, _ = compactness_seg_prob_map(img, probMap, params)

    diceADMM, precisionADMM, recallADMM = eval_results(segADMM, grd_truth)
    diceGCs, precisionGCs, recallGCs = eval_results(segGCs, grd_truth)
    diceCNN, precisionCNN, recallCNN = eval_results(segCNN, grd_truth)

    print(diceADMM, precisionADMM, recallADMM)
    print(diceGCs, precisionGCs, recallGCs)
    print(diceCNN, precisionCNN, recallCNN)

    draw_results(img, grd_truth, segCNN, segGCs, segADMM)
